packages/@aws-cdk/aws-lambda/test/function-hash.test.ts

```diff
@@ -36,103 +37,130 @@ describe('function hash', () => {
     });
 
-    expect(calculateFunctionHash(fn1)).toEqual('d1bc824ac5022b7d62d8b12dbae6580c');
-    expect(calculateFunctionHash(fn2)).toEqual('0f168f0772463e8e547bb3800937e54d');
-  });
+    test('inline code change impacts the hash', () => {
+      const stack1 = new Stack();
+      const fn1 = new lambda.Function(stack1, 'MyFunction', {
+        runtime: lambda.Runtime.NODEJS_12_X,
+        code: lambda.Code.fromInline('foo'),
+        handler: 'index.handler',
+      });
 
-  test('inline code change impacts the hash', () => {
-    const stack1 = new Stack();
-    const fn1 = new lambda.Function(stack1, 'MyFunction', {
-      runtime: lambda.Runtime.NODEJS_12_X,
-      code: lambda.Code.fromInline('foo'),
-      handler: 'index.handler',
-    });
+      const stack2 = new Stack();
+      const fn2 = new lambda.Function(stack2, 'MyFunction', {
+        runtime: lambda.Runtime.NODEJS_10_X,
+        code: lambda.Code.fromInline('foo bar'),
+        handler: 'index.handler',
+      });
 
-    const stack2 = new Stack();
-    const fn2 = new lambda.Function(stack2, 'MyFunction', {
-      runtime: lambda.Runtime.NODEJS_10_X,
-      code: lambda.Code.fromInline('foo bar'),
-      handler: 'index.handler',
+      expect(calculateFunctionHash('v1', fn1)).not.toEqual(calculateFunctionHash('v1', fn2));
     });
 
-    expect(calculateFunctionHash(fn1)).toEqual('ebf2e871fc6a3062e8bdcc5ebe16db3f');
-    expect(calculateFunctionHash(fn2)).toEqual('ffedf6424a18a594a513129dc97bf53c');
-  });
+    describe('impact of env variables order on hash', () => {
+      test('with "currentVersion", we sort env keys so order is consistent', () => {
+        const stack1 = new Stack();
+        const fn1 = new lambda.Function(stack1, 'MyFunction', {
+          runtime: lambda.Runtime.NODEJS_12_X,
+          code: lambda.Code.fromAsset(path.join(__dirname, 'my-lambda-handler')),
+          handler: 'index.handler',
+          environment: {
+            Foo: 'bar',
+            Bar: 'foo',
+          },
+        });
 
-  describe('impact of env variables order on hash', () => {
+        new CfnOutput(stack1, 'VersionArn', { value: fn1.currentVersion.functionArn });
 
-    test('without "currentVersion", we preserve old behavior to avoid unnesesary invalidation of templates', () => {
+        const stack2 = new Stack();
+        const fn2 = new lambda.Function(stack2, 'MyFunction', {
+          runtime: lambda.Runtime.NODEJS_12_X,
+          code: lambda.Code.fromAsset(path.join(__dirname, 'my-lambda-handler')),
+          handler: 'index.handler',
+          environment: {
+            Bar: 'foo',
+            Foo: 'bar',
+          },
+        });
+
+        new CfnOutput(stack2, 'VersionArn', { value: fn2.currentVersion.functionArn });
+
+        expect(calculateFunctionHash('v1', fn1)).toEqual(calculateFunctionHash('v1', fn2));
+      });
+    });
+  });
+
+  describe('calculateFunctionHashV2', () => {
+    test('environment variables impact hash', () => {
```

Comment:
It seems like most of these tests (excluding the `...do not affect function hash`) could be captured by a `describe.each` on the original `calculateFunctionHashV1` function block. Either the one test that explicitly captures the assetHash could be moved outside of that block, or the asset hash could be included as a parameter to the `describe.each`. Would cut down on ~100 lines of copy/pasted test code.

Reply:
The tests are doing two things - (a) check that the hash is equal to or not equal to another function hash (i.e., relative comparison) (b) check that the hash remains constant and does not regress due to an unintended change (i.e., absolute comparison).

By moving to a `describe.each()` block, we'll lose the latter. 
I could write one separate sanity test that does absolute comparison but that would mean reduced coverage.

Reply:
> The tests are doing two things [...] (b) check that the hash remains constant

Yes, that's what I meant by "the one test that explicitly captures the assetHash could be moved outside of that block". This is a nitpick, so feel free to ignore/disagree; it just feels like most of these tests are duplicates.

You can also include the "regression" hashes in the `describe.each` block:

```ts
describe.each([
  ['v1', 'aea5463dba236007afe91d2832b3c836', '979b4a14c6f174c745cdbcd1036cf844'],
  ['v2', 'abc123', 'cba321'], 
])('calculateFunctionHash%s', (version, baseHash, updatedHash) => {
  // All of the comparison/regression tests go here.
});

describe('V2 exclusions', () => {
  // Test for the fields v2 explicitly opts out of
});
```

Reply:
The `describe.each` with the hashes look ugly. They'll make the code harder to read.

I've used a different approach entirely to reduce duplication. I've merged these test cases into one.